daemon/src/engine/framework/Crypto.cpp

```diff
@@ -0,0 +1,289 @@
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL DAEMON DEVELOPERS BE LIABLE FOR ANY
+DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+===========================================================================
+*/
+
+#include "common/Common.h"
+#include "Crypto.h"
+#include <nettle/aes.h>
+#include <nettle/base64.h>
+#include <nettle/md5.h>
+#include <nettle/sha2.h>
+
+// Compatibility with old nettle versions
+#if !defined(AES256_KEY_SIZE)
+
+#define AES256_KEY_SIZE 32
+
+typedef aes_ctx compat_aes256_ctx;
+
+static void nettle_compat_aes256_set_encrypt_key(compat_aes256_ctx *ctx, const uint8_t *key)
+{
+	nettle_aes_set_encrypt_key(ctx, AES256_KEY_SIZE, key);
+}
+
+static void nettle_compat_aes256_set_decrypt_key(compat_aes256_ctx *ctx, const uint8_t *key)
+{
+	nettle_aes_set_decrypt_key(ctx, AES256_KEY_SIZE, key);
+}
+
+static void nettle_compat_aes256_encrypt(const compat_aes256_ctx *ctx,
+						   size_t length, uint8_t *dst,
+						   const uint8_t *src)
+{
+	nettle_aes_encrypt(ctx, length, dst, src);
+}
+
+static void nettle_compat_aes256_decrypt(const compat_aes256_ctx *ctx,
+						   size_t length, uint8_t *dst,
+						   const uint8_t *src)
+{
+	nettle_aes_decrypt(ctx, length, dst, src);
+}
+
+
+static int nettle_compat_base64_decode_update(base64_decode_ctx *ctx,
+		     size_t *dst_length,
+		     uint8_t *dst,
+		     size_t src_length,
+		     const uint8_t *src)
+{
+	unsigned dst_length_uns = *dst_length;
+	return nettle_base64_decode_update(ctx, &dst_length_uns, dst, src_length, src);
+}
+
+#undef aes256_set_encrypt_key
+#define nettle_aes256_set_encrypt_key nettle_compat_aes256_set_encrypt_key
+#define nettle_aes256_set_decrypt_key nettle_compat_aes256_set_decrypt_key
+#define nettle_aes256_invert_key      nettle_compat_aes256_invert_key
+#define nettle_aes256_encrypt         nettle_compat_aes256_encrypt
+#define nettle_aes256_decrypt         nettle_compat_aes256_decrypt
+#define aes256_ctx compat_aes256_ctx
+#define nettle_base64_decode_update nettle_compat_base64_decode_update
+
+#endif // NETTLE_VERSION_MAJOR < 3
+
+namespace Crypto {
+
+Data RandomData( std::size_t bytes )
+{
+    Data data( bytes );
+    Sys::GenRandomBytes(data.data(), data.size());
+    return data;
+}
+
+namespace Encoding {
+
+/*
+ * Translates binary data into a hexadecimal string
+ */
+Data HexEncode( const Data& input )
+{
+    std::ostringstream stream;
```

Comment:
Maybe streams are a bit of an overkill.
It can still be done in a C++ way though.